Feature/celoe refinement

LukasBluebaum opened the pull request:
- Update CELOE refinement operator (refinement steps taken from DL-Learner)
    - Add cardinality restrictions
    - Add boolean, numeric, time datatypes
- As discussed in our last meeting, started to add more methods to the ConceptGenerator to retrieve concepts, properties, property values etc.
    - We can then also think about some caching in the ConceptGenerator (e.g. domain -> property mappings)
- Add AbstractValueSplitter and BinningValueSplitter (used the split calculation from [DL-Learner](https://github.com/SmartDataAnalytics/DL-Learner/blob/e2b5ba8729a8ee2eff412d524fd6e63c6863cd0c/components-core/src/main/java/org/dllearner/utilities/split/AbstractValuesSplitter.java#L88))
- Update ConceptOperandSorter, OperandSetTransform, OrderedOWLObject for data properties

LukasBluebaum commented on ontolearn/refinement_operators.py:
```diff
@@ -244,34 +245,39 @@ def __init__(self,
 
-        props = []
+            self.max_nr_fillers = defaultdict(int)
+            for prop in obj_properties:
```
The calculation of the cardinality limits for each object property can be slow (around 30sec on the daikiri_excerpt dataset on my laptop, on the other datasets it’s around 1-2 sec). Could easily speed this up by using more memory but this would run into problems when using bigger datasets I guess (e.g. the full daikiri dataset). However, it also only runs once at the start and we could also make it optional and only use the specified cardinality limit otherwise.

LukasBluebaum commented on ontolearn/refinement_operators.py:
```diff
@@ -374,7 +384,7 @@ def refine_atomic_concept(self, ce: OWLClass, max_length: Optional[int] = None,
                 max_ = self.max_nr_fillers[prop]
                 if max_ > 1 or (self.use_negation and max_ > 0):
-                    card_res.append(self.kb.max_cardinality_restriction(self.kb.thing, prop, max_-1))
+                    card_res.append(self.kb.max_cardinality_restriction(self.kb.thing, prop, max_ - 1))
```
Why does the refinement start at `max_filler - 1` here in the [DL-Learner implementation](https://github.com/SmartDataAnalytics/DL-Learner/blob/e2b5ba8729a8ee2eff412d524fd6e63c6863cd0c/components-core/src/main/java/org/dllearner/refinementoperators/RhoDRDown.java#L1403)? Should it not start at the `max_filler`?

LukasBluebaum commented on ontolearn/value_splitter.py:
```diff
@@ -21,61 +18,45 @@ def __init__(self, max_nr_splits: int):
-        if len(values) > 1:
-            splits.add(values[len(values)-1])
-
         for i in range(1, nr_splits):
```
More in general: I guess having code that’s really close to the DL-Learner code is no problem since Ontolearn is also GPL licensed?

LukasBluebaum commented on ontolearn/refinement_operators.py:
```diff
@@ -562,29 +565,28 @@ def refine_data_some_values_from(self, ce: OWLDataSomeValuesFrom) -> Iterable[OW
             splits = self.dp_splits[ce.get_property()]
-            for facet_res in datarange.get_facet_restrictions():
+            if len(splits) > 0 and len(datarange.get_facet_restrictions()) > 0:
+                facet_res = datarange.get_facet_restrictions()[0]
```
In general, the refinement operator will just add DatatypeRestrictions with only one facet restriction. The [DL-Learner implementation](https://github.com/SmartDataAnalytics/DL-Learner/blob/e2b5ba8729a8ee2eff412d524fd6e63c6863cd0c/components-core/src/main/java/org/dllearner/refinementoperators/RhoDRDown.java#L695) loops through the FacetRestrictions and takes the last one, if there are somehow more than one. That should not happen or is there something I’m missing here? I just take the first one, we could also throw an error or do nothing if there is more than one but I think this is fine.

LukasBluebaum reviewed (commented):
Ignore this comment ;D

LukasBluebaum commented on ontolearn/refinement_operators.py:
```diff
@@ -244,34 +245,39 @@ def __init__(self,
 
-        props = []
+            self.max_nr_fillers = defaultdict(int)
+            for prop in obj_properties:
+                for ind in self.kb.individuals():
+                    num = sum(1 for _ in zip(self.kb.get_object_property_values(ind, prop), range(self.card_limit)))
+                    self.max_nr_fillers[prop] = max(self.max_nr_fillers[prop], num)
+                    if num == self.card_limit:
+                        break
```
The calculation of the cardinality limits for each object property can be slow (around 30sec on the daikiri_excerpt dataset on my laptop, on the other datasets it’s around 1-2 sec). Could easily speed this up by using more memory but this would run into problems when using bigger datasets I guess (e.g. the full daikiri dataset). However, it also only runs once at the start and we could also make it optional and only use the specified cardinality limit otherwise.

Demirrr replied:
I reckon that there is some room to improve this bit of code 
`num = sum(1 for _ in zip(self.kb.get_object_property_values(ind, prop), range(self.card_limit)))`

1. What is the need for `range(self.card_limit))`?
2. Is the output of `self.kb.get_object_property_values(ind, prop)` generator ?

LukasBluebaum replied:
1. `self.card_limit` is the static cardinality limit. So even if there is an individual which is connected to more objects (individuals) via a property, `self.card_limit` is used as the `max_nr_fillers` for that property. The line then counts the number of elements in the generator until it is exhausted or `self.card_limit` is reached.
2. Yes

The main problem here should be that for each property the access to a generator is done for each individual. It should be way faster to do this in one pass over the triples/relations of each property (e.g. via `get_relations()` from owlready2) and then count the number of objects for each subject and take the max at the end. That would use more memory though, so I´m not sure if that could lead to problems when using bigger datasets. It could also just be optional I guess and otherwise, we just use the `self.card_limit` for each property and skip this computation.

Demirrr replied:
Thank you for your explanation. Let me revise my previous comment by stating that it is not very obvious anymore to me how to improve the aforementioned bit.

LukasBluebaum commented on ontolearn/refinement_operators.py:
```diff
@@ -374,7 +384,7 @@ def refine_atomic_concept(self, ce: OWLClass, max_length: Optional[int] = None,
                 max_ = self.max_nr_fillers[prop]
                 if max_ > 1 or (self.use_negation and max_ > 0):
-                    card_res.append(self.kb.max_cardinality_restriction(self.kb.thing, prop, max_-1))
+                    card_res.append(self.kb.max_cardinality_restriction(self.kb.thing, prop, max_ - 1))
```
Why does the refinement start at `max_filler - 1` here in the [DL-Learner implementation](https://github.com/SmartDataAnalytics/DL-Learner/blob/e2b5ba8729a8ee2eff412d524fd6e63c6863cd0c/components-core/src/main/java/org/dllearner/refinementoperators/RhoDRDown.java#L1403)? Should it not start at the `max_filler`?

SimonBin replied:
According to the docs, the initial refinements that DL-Learner builds are put into "buckets" mAx, x=length of the construct. 

>=maxFillers is taken to have the same closed world interpretation as forAll

So it puts  <=maxFillters-1 into the initial refinement set.

To the best I can tell, this is neither applied consistently (>= and = are raised up to maxFillers) nor can I see any obvious consequences of doing it this way or the other way...

LukasBluebaum commented on ontolearn/value_splitter.py:
```diff
@@ -21,61 +18,45 @@ def __init__(self, max_nr_splits: int):
         for i in range(1, nr_splits):
             index = max(math.floor(i * len(values) / nr_splits),
                         math.floor(i * len(values) / (nr_splits - 1) - 1))
             splits.add(self._combine_values(values[index], values[min(index + 1, len(values)-1)]))
```
More in general: I guess having code that’s really close to the DL-Learner code is no problem since Ontolearn is also GPL licensed?

LukasBluebaum commented on ontolearn/refinement_operators.py:
```diff
@@ -562,29 +565,28 @@ def refine_data_some_values_from(self, ce: OWLDataSomeValuesFrom) -> Iterable[OW
+                facet_res = datarange.get_facet_restrictions()[0]
                 val = facet_res.get_facet_value()
-                index = splits.index(val)
+                idx = splits.index(val)
```
In general, the refinement operator will just add DatatypeRestrictions with only one facet restriction. The [DL-Learner implementation](https://github.com/SmartDataAnalytics/DL-Learner/blob/e2b5ba8729a8ee2eff412d524fd6e63c6863cd0c/components-core/src/main/java/org/dllearner/refinementoperators/RhoDRDown.java#L695) loops through the FacetRestrictions and takes the last one, if there are somehow more than one. That should not happen or is there something I’m missing here? I just take the first one, we could also throw an error or do nothing if there is more than one but I think this is fine.

SimonBin replied:
good observation, in theory we should be able to have restrictions like `someDecimal[<=2.0, >= 1.0]` which consist of 2 facets but it doesn't seem like DL-Learner handles it

LukasBluebaum reviewed (commented):
Some comments/question below.

Demirrr commented:
Hello @LukasBluebaum,

Thank you for your implementations. I was wondering whether you have also written tests for these changes.

LukasBluebaum commented:
Hi @Demirrr, I added a [test](https://github.com/dice-group/Ontolearn/blob/1665c435f0142259ca6763cc6703af00b70a4a0b/tests/test_celoe.py#L63) for CELOE for mutagenesis, so basically a regression test as the other ones in that file. And I added a test for the new ValueSplitter. However, the [refinement operators](https://github.com/dice-group/Ontolearn/blob/develop/tests/test_refinement_operators.py) and the [ConceptGenerator](https://github.com/dice-group/Ontolearn/blob/develop/tests/test_concept_generator.py) still have no tests. The refinements have some but they just run the operator and don't check anything. I can start to add some tests for these too.

Demirrr commented:
> Hi @Demirrr, I added a [test](https://github.com/dice-group/Ontolearn/blob/1665c435f0142259ca6763cc6703af00b70a4a0b/tests/test_celoe.py#L63) for CELOE for mutagenesis, so basically a regression test as the other ones in that file. And I added a test for the new ValueSplitter. However, the [refinement operators](https://github.com/dice-group/Ontolearn/blob/develop/tests/test_refinement_operators.py) and the [ConceptGenerator](https://github.com/dice-group/Ontolearn/blob/develop/tests/test_concept_generator.py) still have no tests. The refinements have some but they just run the operator and don't check anything. I can start to add some tests for these too.

Thank you for adding a test and the update! 

If your workload allows, it would be quite good to add few tests.

SimonBin commented on owlapy/model/__init__.py:
```diff
@@ -1245,6 +1259,19 @@ def data_property_values(self, ind: OWLNamedIndividual, pe: OWLDataProperty) ->
         pass
 
+    @abstractmethod
+    def all_data_property_values(self, pe: OWLDataProperty) -> Iterable['OWLLiteral']:
```
this one is the most tricky part of your suggestion, as it is not part of the base reasoner requirements and it comes a bit in the way when we want to stick very close to OWLAPI.... 

a) take this suggestion, break compat with OWLAPI
b) migrate our code to some "OWLReasonerExtended" that wraps the OWLReasoner and provides extra functionality (DL-Learner did something like this)
c) add a static helper function which special-cases this

SimonBin replied:
(also note that the current implementation seems to yield duplicate values, i.e. no set behaviour)
